Add `--appMetadataFile` deploy flag

DennisDenuto opened the pull request:
TODO: 
- [x] remove any hardcoding left over
- [x] fix up commits

#### What this PR does / why we need it:
related to: https://github.com/vmware-tanzu/carvel-kapp-controller/pull/799

#### Does this PR introduce a user-facing change?
Yes, it introduces a new flag to write the app metadata config map to disk

```release-note
- Adds deploy flag `--appMetadataFile` to write an app metadata file to disk
```

#### Additional Notes for your reviewer:

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

praveenrewar commented on pkg/kapp/cmd/app/deploy_flags.go:
```diff
@@ -48,4 +49,5 @@ func (s *DeployFlags) Set(cmd *cobra.Command) {
 
 	cmd.Flags().BoolVar(&s.Logs, "logs", true, fmt.Sprintf("Show logs from Pods annotated as '%s'", deployLogsAnnKey))
 	cmd.Flags().BoolVar(&s.LogsAll, "logs-all", false, "Show logs from all Pods")
+	cmd.Flags().StringVar(&s.AppMetadataFile, "appMetadataFile", "", "write app metatadata to file")
```
`app-metadata-file`?

DennisDenuto replied:
yeah good suggestion! i'll change it now

praveenrewar reviewed (commented):
> a new flag to write the app metadata config map to disk

Nice!! I am curious to know the use case for this :)

*Update*: Never mind, just noticed https://github.com/vmware-tanzu/carvel-kapp-controller/pull/799

DennisDenuto committed:
feat: Add --app-metadata-file flag

Saves app configmap metadata to disk allowing consumers to discover the namespaces and GK's of the deployed app

praveenrewar commented on pkg/kapp/cmd/app/deploy.go:
```diff
@@ -225,8 +227,25 @@ func (o *DeployOptions) Run() error {
+			if err != nil {
+				return err
+			}
+			err = ioutil.WriteFile(o.DeployFlags.AppMetadataFile, []byte(meta.AsString()), os.ModePerm)
```
Maybe we should directly use `os.WriteFile()`? (Since that's just what `ioutil.WriteFile()` is doing)
Similar in e2e test.

praveenrewar commented on pkg/kapp/cmd/app/deploy.go:
```diff
@@ -225,8 +227,25 @@ func (o *DeployOptions) Run() error {
+				return err
+			}
+			err = ioutil.WriteFile(o.DeployFlags.AppMetadataFile, []byte(meta.AsString()), os.ModePerm)
+			if err != nil {
+				return err
+			}
```
Should we just return the err at the end instead of nil? (No strong opinions here as handling the err more closer to it’s  occurrence seems more readable.)

DennisDenuto replied:
sgtm! returned an err instead of nil at the end

praveenrewar replied:
hehe, I actually meant that we can remove the err check and just return err at the end instead of nil. 
```go
			err = os.WriteFile(o.DeployFlags.AppMetadataFile, []byte(meta.AsString()), os.ModePerm)
		}
		return err
```
But I am okay with the current way as well because err should be handled closer to it’s  origin.

praveenrewar commented on pkg/kapp/cmd/app/deploy.go:
```diff
@@ -225,8 +227,25 @@ func (o *DeployOptions) Run() error {
+		}
+
+		return nil
+
```
nit: remove newline

praveenrewar commented on test/e2e/app_change_test.go:
```diff
@@ -89,3 +93,112 @@ spec:
+	cleanUp()
+	defer cleanUp()
+
+	firstDeploy, err := ioutil.TempFile(os.TempDir(), "output1")
```
Same as above, we could maybe use `os.CreateTemp` instead?

praveenrewar commented on pkg/kapp/cmd/app/deploy_flags.go:
```diff
@@ -48,4 +49,5 @@ func (s *DeployFlags) Set(cmd *cobra.Command) {
 
 	cmd.Flags().BoolVar(&s.Logs, "logs", true, fmt.Sprintf("Show logs from Pods annotated as '%s'", deployLogsAnnKey))
 	cmd.Flags().BoolVar(&s.LogsAll, "logs-all", false, "Show logs from all Pods")
+	cmd.Flags().StringVar(&s.AppMetadataFile, "app-metadata-file", "", "write app metatadata to file")
```
nit:
```suggestion
	cmd.Flags().StringVar(&s.AppMetadataFile, "app-metadata-file", "", "Write app metadata to file")
```

Also, maybe we would want to suggest that it accepts a filename as input? Something like "Set filename to write app metadata"?

100mik commented on test/e2e/app_change_test.go:
```diff
@@ -89,3 +93,112 @@ spec:
+	logger := Logger{}
+	kapp := Kapp{t, env.Namespace, env.KappBinaryPath, logger}
+
+	yaml1 := `
```
Not a blocker per say, but in interest of keeping the tests minimal 🤔 

We could have a ConfigMap and a Secret instead of a Deployment and a Service right? 
I think it would be less verbose and probably make the test faster

praveenrewar commented on test/e2e/app_change_test.go:
```diff
@@ -89,3 +93,86 @@ spec:
+		kapp.RunWithOpts([]string{"deploy", "--app-metadata-file", secondDeploy.Name(), "-f", "-", "-a", name}, RunOpts{IntoNs: true, StdinReader: strings.NewReader(yaml2)})
+	})
+
+	configMapFirstDeploy, err := ioutil.ReadFile(firstDeploy.Name())
```
(I think I forgot to add a comment about this one and at line 158)
*Suggestion*: Use `os.ReadFile()` instead

praveenrewar reviewed (approved):
LGTM!

100mik reviewed (approved):
Looks good!

cppforlife commented on test/e2e/app_change_test.go:
```diff
@@ -5,6 +5,9 @@ package e2e
 
 import (
 	"fmt"
+	"github.com/stretchr/testify/assert"
+	"gopkg.in/yaml.v2"
```
nit: external imports in separate section

cppforlife commented on pkg/kapp/cmd/app/deploy_flags.go:
```diff
@@ -48,4 +49,5 @@ func (s *DeployFlags) Set(cmd *cobra.Command) {
 
 	cmd.Flags().BoolVar(&s.Logs, "logs", true, fmt.Sprintf("Show logs from Pods annotated as '%s'", deployLogsAnnKey))
 	cmd.Flags().BoolVar(&s.LogsAll, "logs-all", false, "Show logs from all Pods")
+	cmd.Flags().StringVar(&s.AppMetadataFile, "app-metadata-file", "", "Set filename to write app metadata")
```
@praveenrewar @100mik is the plan here would also be to read from such file in future? if not, we should probably follow carvel convention of having `-output` suffix.

praveenrewar replied:
> is the plan here would also be to read from such file in future?

I can't think of any reason we would want to read from the file instead of the ConfigMap. So yeah, we should probably add the `-output` suffix.

DennisDenuto replied:
sgtm; added `-output` suffix to the flag name

sethiyash reviewed (approved):
LGTM!
